packages/upload-api/src/blob/allocate.js

```diff
@@ -1,95 +1,129 @@
-          },
-        }
-      }
+/**
+ * @param {API.BlobServiceContext} context
+ * @param {API.ProviderInput<API.BlobAllocate>} input
+ */
+export const allocate = async (context, { capability }) => {
+  // Only service principal can perform an allocation
+  if (capability.with !== context.id.did()) {
+    return {
+      error: new UnsupportedCapability({ capability }),
+    }
+  }
 
-      // Allocate memory space for the blob. If memory for this blob is
-      // already allocated, this allocates 0 bytes.
-      const allocationInsert = await context.allocationsStorage.insert({
-        space,
-        blob,
-        cause,
-      })
-      if (allocationInsert.error) {
-        // if the insert failed with conflict then this item has already been
-        // added to the space and there is no allocation change.
-        // If record exists but is expired, it can be re-written
-        if (allocationInsert.error.name === 'RecordKeyConflict') {
-          size = 0
-        } else {
-          return {
-            error: allocationInsert.error,
-          }
-        }
-      }
+  const { blob, cause, space } = capability.nb
+  let { size } = blob
 
-      // Check if blob already exists
-      // TODO: this may depend on the region we want to allocate and will need changes in the future.
-      const hasBlobStore = await context.blobsStorage.has(blob.digest)
-      if (hasBlobStore.error) {
-        return hasBlobStore
-      }
+  // We check if space has storage provider associated. If it does not
+  // we return `InsufficientStorage` error as storage capacity is considered
+  // to be 0.
+  const result = await context.provisionsStorage.hasStorageProvider(space)
+  if (result.error) {
+    return result
+  }
 
-      // If blob is stored, we can just allocate it to the space with the allocated size
-      // TODO: this code path MAY lead to await failures - awaited http/put and blob/accept tasks
-      // are supposed to fail if path does not exists.
-      if (hasBlobStore.ok) {
-        return {
-          ok: { size },
-        }
-      }
+  if (!result.ok) {
+    return {
+      /** @type {API.AllocationError} */
+      error: {
+        name: 'InsufficientStorage',
+        message: `${space} has no storage provider`,
+      },
+    }
+  }
 
-      // Get presigned URL for the write target
-      const expiresIn = 60 * 60 * 24 // 1 day
-      const expiresAt = new Date(Date.now() + expiresIn).toISOString()
-      const createUploadUrl = await context.blobsStorage.createUploadUrl(
-        blob.digest,
-        blob.size,
-        expiresIn
-      )
-      if (createUploadUrl.error) {
-        return createUploadUrl
+  // Verify blob is within the max upload size.
+  if (capability.nb.blob.size > context.maxUploadSize) {
+    // While blob may exceed current maxUploadSize limit it could be that limit
+    // was higher in the past and user had this blob uploaded already in which
+    // case we should not error.
```

Comment:
I am not sure why this moved here instead of `blob/add`. Would you expect that write targets could have different max sizes? Otherwise, I don't understand the change as it results in way more processing for same end result

Reply:
I think you are making some good points on this. That said, I also think that I also think there should be no `blob/add` handler it should be a static workflow like one illustrated here https://github.com/w3s-project/RFC/pull/25#issuecomment-2070641427

Perhaps what we miss is a prior step that checks that blob size is below the max limit (and that step can even run locally on the invoking agent (like http/put) as opposed to service.

Overall sounds like spec needs to address this by calling out the verification step that needs to take place. In the meantime I suggest we keep this here and we can address this as part of router design which will actually deal with multiple allocation sites. Does this sound reasonable ?

If you feel very strongly about it I can move this back into the blob/add handler, however please note that allocation site would need to have this kind of check anyway as they may have different constraints to work with.